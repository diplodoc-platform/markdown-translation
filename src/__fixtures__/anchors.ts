const markdown = `\
# заголовок без явного якоря
параграф

## заголовок с явным якорем {#anchor}
`;

const skeleton = `
# %%%1%%%
%%%2%%%

## %%%3%%% {#anchor}
`;

const translations = new Map<string, string>([
    ['1', 'heading with implicit anchor'],
    ['2', 'paragraph'],
    ['3', 'heading with explicit anchor'],
]);

export {markdown, skeleton, translations};
export default {markdown, skeleton, translations};
