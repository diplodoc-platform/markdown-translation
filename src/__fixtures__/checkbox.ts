const markdown = `
[ ] первый

[x] второй

- [_] надо сделать подчёркивание
- [-] надо сделать тире
- [ ] надо сделать пробел
- [x] сделано крестик
- [X] сделано крестик заглавный

- элемент листа
  - [x] сделан
  - [ ] надо сделaть
`;

const skeleton = `\
[ ] %%%1%%%

[x] %%%2%%%
- [_] %%%3%%%
- [-] %%%4%%%
- [ ] %%%5%%%
- [x] %%%6%%%
- [X] %%%7%%%

- %%%8%%%
  - [x] %%%9%%%
  - [ ] %%%10%%%
`;

const translations = new Map<string, string>([
    ['1', 'first'],
    ['2', 'second'],
    ['3', 'to be done underscore'],
    ['4', 'to be done dash'],
    ['5', 'to be done space'],
    ['6', 'done cross'],
    ['7', 'done capital cross'],
    ['8', 'list element'],
    ['9', 'done'],
    ['10', 'to be done'],
]);

export {markdown, skeleton, translations};
export default {markdown, skeleton, translations};
