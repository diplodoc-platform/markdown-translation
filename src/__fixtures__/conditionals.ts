const markdown = `# Document Title

{% if OS == 'iOS' %}

Download the app from the [{{ ios-marketplace }} Store](https://www.apple.com/ios/app-store/).

{% else %}

Download the app from [{{ other-marketplace }} Store](https://play.google.com).

Some text {% if  OS == 'iOS' %} Apple {% else %} Android {% endif %} text continued.

{% endif %}
`;

const skeleton = `# %%%1%%%

{% if OS == 'iOS' %}

%%%2%%% [%%%3%%%](https://www.apple.com/ios/app-store)%%%4%%%

{% else %}

%%%5%%% [%%%6%%%](https://play.google.com)%%%7%%%

%%%8%%% {% if  OS == 'iOS' %} %%%9%%% {% else %} %%%10%%% {% endif %} %%%11%%%

{% endif %}
`;

const translations = new Map<string, string>([
    ['1', 'Титул Документа'],
    ['2', 'Скачайте приложение из '],
    ['3', '{{ ios-marketplace }} Стора'],
    ['4', '.'],
    ['5', 'Скачайте приложение из '],
    ['6', '{{ other-marketplace }} Стора'],
    ['7', '.'],
    ['8', 'Какой-то текст'],
    ['9', 'Apple'],
    ['10', 'Android'],
    ['11', 'Продолжение текста.'],
]);

export {markdown, skeleton, translations};
export default {markdown, skeleton, translations};
