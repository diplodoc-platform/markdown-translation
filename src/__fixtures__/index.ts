const markdown = `# Заголовок 1

Параграф

- Первый элемент списка
- Второй элемент списка

Заголовок 2
-----------

Параграф содержащий **жирный текст** и *курсив*

> текст внутри цитаты

1. Первый элемент нумерованного списка
2. Второй элемент нумерованного списка

[Текст ссылки](files/file.md "Описание ссылки")

![Альтернативный текст картинки](files/image.png "Титул картинки")
`;

const skeleton = `# %%%1%%%
%%%2%%%
- %%%3%%%
- %%%4%%%
%%%5%%%
-----------
%%%6%%%**%%%7%%%**%%%8%%%*%%%9%%%*
> %%%10%%%
1. %%%11%%%
2. %%%12%%%

[%%%13%%%](files/file.md "%%%14%%%")
`;

const translations = new Map<string, string>([
    ['1', 'Heading 1'],
    ['2', 'Paragraph'],
    ['3', 'First list item'],
    ['4', 'Second list item'],
    ['5', 'Heading 2'],
    ['6', 'Paragraph with '],
    ['7', 'bold text'],
    ['8', ' and '],
    ['9', 'italic'],
    ['10', 'text inside of the blockquote'],
    ['11', 'First ordered list item'],
    ['12', 'Second ordered list item'],
    ['13', 'Link text'],
    ['14', 'Link Description'],
]);

export {markdown, skeleton, translations};
export default {markdown, skeleton, translations};
