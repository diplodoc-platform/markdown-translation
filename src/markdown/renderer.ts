import MarkdownIt from 'markdown-it';
import {
    mdRenderer,
    MarkdownRendererParams,
    MarkdownRendererEnv,
} from '@diplodoc/markdown-it-markdown-renderer';

// configure with diplodoc plugins
// @ts-ignore
import meta from 'markdown-it-meta';
import notes from '@doc-tools/transform/lib/plugins/notes';

import markdownHandlers, {MarkdownHandlersState} from './handlers';
import hooks, {HooksParameters} from './hooks';

export type MarkdownRendererState = MarkdownHandlersState;

export type RenderParameters = {
    skeleton: string;
    translations: Map<string, string>;
} & DiplodocParameters;

export type DiplodocParameters = {
    lang?: string;
};

function render(parameters: RenderParameters) {
    if (!validParameters(parameters)) {
        throw new Error('invalid parameters');
    }

    const md = new MarkdownIt('commonmark', {html: true}) as HooksParameters['markdownit'];

    const {handlers, initState} = markdownHandlers.generate(parameters);
    const markdownHooks = hooks.generate({markdownit: md});
    const mdOptions: MarkdownRendererParams<MarkdownHandlersState> = {
        handlers,
        initState,
        hooks: markdownHooks.hooks,
    };
    const diplodocOptions = {
        lang: parameters.lang ?? 'ru',
    };

    const env: MarkdownRendererEnv = {
        source: parameters.skeleton.split('\n'),
    };

    md.use(mdRenderer, mdOptions);

    // diplodoc plugins
    md.use(meta);
    md.use(notes, diplodocOptions);

    return md.render(parameters.skeleton, env);
}

function validParameters(parameters: RenderParameters) {
    const {skeleton, translations} = parameters;

    const skeletonCondition = skeleton !== undefined;
    const translationsCondition = translations !== undefined;

    const conditions = [
        skeletonCondition,
        translationsCondition,
        validDiplodocParameters(parameters),
    ];

    return conditions.reduce((a, v) => a && v, true);
}

function validDiplodocParameters(parameters: DiplodocParameters) {
    const {lang} = parameters;

    return lang === undefined || lang === 'ru' || lang === 'en';
}

export {render, validParameters, validDiplodocParameters};
export default {render, validParameters, validDiplodocParameters};
