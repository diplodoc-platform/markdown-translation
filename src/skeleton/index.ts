export * as renderer from './renderer';
export * as default from './index';
