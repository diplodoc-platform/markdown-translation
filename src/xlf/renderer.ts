import MarkdownIt from 'markdown-it';
import {customRenderer, CustomRendererParams} from '@diplodoc/markdown-it-custom-renderer';

// configure with diplodoc plugins
// @ts-ignore
import meta from 'markdown-it-meta';
import notes from '@doc-tools/transform/lib/plugins/notes';

import {template} from './generator';
import rules, {XLFRulesState} from './rules';
import hooks, {HooksParameters} from './hooks';

export type XLFRendererState = XLFRulesState;

export type RenderParameters = {
    markdown: string;
} & template.TemplateParameters &
    DiplodocParameters;

export type DiplodocParameters = {
    lang?: string;
};

function render(parameters: RenderParameters) {
    if (!validParameters(parameters)) {
        throw new Error('invalid parameters');
    }

    const wrapper = template.generate(parameters);

    const xlfRenderer = new MarkdownIt('commonmark', {html: true}) as HooksParameters['markdownit'];
    const xlfRules = rules.generate(wrapper);
    const xlfHooks = hooks.generate({template: wrapper.template, markdownit: xlfRenderer});
    const xlfOptions: CustomRendererParams<XLFRendererState> = {
        rules: xlfRules.rules,
        hooks: xlfHooks.hooks,
        initState: xlfRules.initState,
    };
    const diplodocOptions = {
        lang: parameters.lang ?? 'ru',
    };

    xlfRenderer.use(customRenderer, xlfOptions);

    // diplodoc plugins
    xlfRenderer.use(meta);
    xlfRenderer.use(notes, diplodocOptions);

    return xlfRenderer.render(parameters.markdown);
}

function validParameters(parameters: RenderParameters) {
    const {lang, markdown} = parameters;

    const markdownCondition = markdown !== undefined;
    const langCondition = lang === undefined || lang === 'ru' || lang === 'en';
    const conditions = [template.validParameters(parameters), markdownCondition, langCondition];

    return conditions.reduce((a, v) => a && v, true);
}

export {render, validParameters};
export default {render, validParameters};
