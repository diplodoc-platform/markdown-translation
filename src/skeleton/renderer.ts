import MarkdownIt from 'markdown-it';
import {
    mdRenderer,
    MarkdownRendererParams,
    MarkdownRendererEnv,
} from '@diplodoc/markdown-it-markdown-renderer';

// configure with diplodoc plugins
// @ts-ignore
import meta from 'markdown-it-meta';
// @ts-ignore
import sup from 'markdown-it-sup';
import notes from '@doc-tools/transform/lib/plugins/notes';
import cut from '@doc-tools/transform/lib/plugins/cut';
import checkbox from '@doc-tools/transform/lib/plugins/checkbox';

import skeletonHandlers, {SkeletonHandlersState} from './handlers';
import hooks, {HooksParameters} from './hooks';

export type SkeletonRendererState = SkeletonHandlersState;

export type RenderParameters = {
    markdown: string;
} & DiplodocParameters;

export type DiplodocParameters = {
    lang?: string;
};

function render(parameters: RenderParameters) {
    if (!validParameters(parameters)) {
        throw new Error('invalid parameters');
    }

    const {markdown, lang} = parameters;

    const md = new MarkdownIt({html: true}) as HooksParameters['markdownit'];
    const env: MarkdownRendererEnv = {source: markdown.split('\n')};

    const {handlers, initState} = skeletonHandlers.generate();

    const skeletonHooks = hooks.generate({markdownit: md});
    const mdOptions: MarkdownRendererParams<SkeletonHandlersState> = {
        handlers,
        initState,
        hooks: skeletonHooks.hooks,
    };
    const diplodocOptions = {
        lang: lang ?? 'ru',
    };

    md.use(mdRenderer, mdOptions);

    // diplodoc plugins
    md.use(meta, diplodocOptions);
    md.use(notes, diplodocOptions);
    md.use(cut, diplodocOptions);
    md.use(sup, diplodocOptions);
    md.use(checkbox, diplodocOptions);

    return md.render(markdown, env);
}

function validParameters(parameters: RenderParameters) {
    const {markdown, lang} = parameters;

    const markdownCondition = markdown !== undefined;
    const langCondition = lang === undefined || lang === 'ru' || lang === 'en';
    const conditions = [markdownCondition, langCondition];

    return conditions.reduce((a, v) => a && v, true);
}

export {render, validParameters};
export default {render, validParameters};
