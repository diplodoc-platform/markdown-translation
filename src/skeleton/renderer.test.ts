import {render, RenderParameters} from './renderer';
import {text} from 'src/__fixtures__';

describe('smoke', () => {
    test('it works', () => {
        const parameters = {
            markdown: '1',
        };

        render(parameters);
    });
});

describe('validates parameters', () => {
    it('works with valid parameters', () => {
        const parameters = {
            markdown: '1',
        };

        render(parameters);
    });

    it('throws on invalid parameters', () => {
        const parameters = {} as RenderParameters;

        expect(() => render(parameters)).toThrow();
    });
});

describe('skeleton rendering', () => {
    it('renders hashes instead of content of the text tokens', () => {
        const parameters = {
            markdown: text,
        };

        const rendered = render(parameters);
        expect(rendered).toMatchSnapshot();
    });
});
