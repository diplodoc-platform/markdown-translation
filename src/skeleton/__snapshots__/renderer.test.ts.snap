// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`skeleton rendering renders hashes instead of content of the text tokens 1`] = `
"# %%%1%%%
%%%2%%%
- %%%3%%%
- %%%4%%%
%%%5%%%
-----------
%%%6%%% **%%%7%%%** %%%8%%% *%%%9%%%*

%%%10%%% %%%11%%%
> %%%12%%%
1. %%%13%%%
2. %%%14%%%

[%%%15%%% %%%16%%%](files/file.md \\"%%%17%%% %%%18%%%\\")

![%%%19%%% %%%20%%%](files/image.png \\"%%%21%%% %%%22%%%\\")"
`;
